### PR TITLE
Add support for the product configuration files

### PR DESCRIPTION
**Pull request by poncovka**
There are new configuration files for all supported products at
`/etc/anaconda/product.d`. The configuration files specify only
a product and a base product for now.

The `ProductLoader` is a class for loading products from configuration
files. It provides access to loaded products, resolves dependencies
between products and collects configuration files for the specified
product.

### CODE REVIEW

File: pyanaconda/core/configuration/product.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+ProductKey.__doc__ = "Identification of the product."
+ProductKey.__str__ = lambda key: " ".join(filter(None, key))
+
+
+ProductData = namedtuple("ProductData", ["base_product", "config_path"])
+ProductData.__doc__ = "Data of the product."
+
+
+class ProductLoader(object):
+    """A class for loading products from configuration files."""
```
[the remark this answers is not in the document]
Comment (poncovka) [reply pyanaconda/core/configuration/product.py]: Fixed.

File: pyanaconda/core/configuration/product.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+    @classmethod
+    def from_defaults(cls):
+        """Create a product loader with default products.
```
[the remark this answers is not in the document]
Comment (poncovka) [reply pyanaconda/core/configuration/product.py]: Fixed.

File: pyanaconda/core/configuration/product.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+        :return: an instance of ProductLoader
+        """
+        loader = cls()
+        loader.load_products(os.path.join(ANACONDA_CONFIG_DIR, "product.d"))
+        return loader
+
+    def __init__(self):
+        """Create a new loader of products."""
```
[the remark this answers is not in the document]
Comment (poncovka) [reply pyanaconda/core/configuration/product.py]: Fixed.

File: pyanaconda/core/configuration/product.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+        self._products = {}
+
+    def load_products(self, config_dir):
+        """Load products from the given configuration directory.
```
[the remark this answers is not in the document]
Comment (poncovka) [reply pyanaconda/core/configuration/product.py]: Fixed.

File: pyanaconda/core/configuration/product.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+            config_path = os.path.join(config_dir, file_name)
+
+            try:
+                self.load_product(config_path)
+            except ConfigurationError as e:
+                log.warning("Skipping an invalid configuration at %s: %s", config_path, e)
+
+    def load_product(self, config_path):
+        """Load a product from the given configuration file.
```
[the remark this answers is not in the document]
Comment (poncovka) [reply pyanaconda/core/configuration/product.py]: Fixed.

File: pyanaconda/core/configuration/product.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+        for file_name in os.listdir(config_dir):
+            if not file_name.endswith(".conf"):
+                continue
+
+            config_path = os.path.join(config_dir, file_name)
+
+            try:
+                self.load_product(config_path)
+            except ConfigurationError as e:
+                log.warning("Skipping an invalid configuration at %s: %s", config_path, e)
```
[the remark this answers is not in the document]
Comment (poncovka) [reply pyanaconda/core/configuration/product.py]: Yes.

File: tests/nosetests/pyanaconda_tests/product_test.py [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+
+        with self.assertRaises(ConfigurationError):
+            self._load_product(content)
+
+        content = dedent("""
+        [Product]
+        variant_name = Server
+
+        """)
+
```
Comment (rvykydal) [inline tests/nosetests/pyanaconda_tests/product_test.py]: This content seems to be the same as the content above ?

Comment (poncovka) [reply tests/nosetests/pyanaconda_tests/product_test.py]: Yes, it is. Fixed.

File: pyanaconda/core/configuration/product.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jkonecny12) [reply pyanaconda/core/configuration/product.py]: The point is to find that immediately in the logs so to make easier debugging. There could be reasons that the fail would be other than we expect and this is an error so or so.

[the remark this answers is not in the document]
Comment (jkonecny12) [reply pyanaconda/core/configuration/product.py]: No, I mean `log.error` :)

File: pyanaconda/core/configuration/product.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+        try:
+            self._get_product_bases(product_key)
+        except ConfigurationError as e:
+            log.warning("The product %s is not valid: %s", product_key, e)
+            return False
+
+        return True
+
+    def collect_configurations(self, product_name, variant_name=""):
+        """Collect configuration files of the the given product.
```
[the remark this answers is not in the document]
Comment (poncovka) [reply pyanaconda/core/configuration/product.py]: Fixed.

File: pyanaconda/core/configuration/product.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (poncovka) [reply pyanaconda/core/configuration/product.py]: Fixed.

**Pull request merged**